Lib9c/Action/BattleArena.cs

```diff
@@ -0,0 +1,333 @@
+            }
+
+            if (!avatarState.worldInformation.TryGetUnlockedWorldByStageClearedBlockIndex(
+                    out var world))
+            {
+                throw new NotEnoughClearedStageLevelException(
+                    $"{addressesHex}Aborted as NotEnoughClearedStageLevelException");
+            }
+
+            if (world.StageClearedId < GameConfig.RequireClearedStageLevel.ActionsInRankingBoard)
+            {
+                throw new NotEnoughClearedStageLevelException(
+                    addressesHex,
+                    GameConfig.RequireClearedStageLevel.ActionsInRankingBoard,
+                    world.StageClearedId);
+            }
+
+            var sheets = states.GetSheets(
+                containArenaSimulatorSheets: true,
+                sheetTypes: new[]
+                {
+                    typeof(ArenaSheet),
+                    typeof(ItemRequirementSheet),
+                    typeof(EquipmentItemRecipeSheet),
+                    typeof(EquipmentItemSubRecipeSheetV2),
+                    typeof(EquipmentItemOptionSheet),
+                    typeof(MaterialItemSheet),
+                });
+
+            avatarState.ValidEquipmentAndCostume(costumes, equipments,
+                sheets.GetSheet<ItemRequirementSheet>(),
+                sheets.GetSheet<EquipmentItemRecipeSheet>(),
+                sheets.GetSheet<EquipmentItemSubRecipeSheetV2>(),
+                sheets.GetSheet<EquipmentItemOptionSheet>(),
+                context.BlockIndex, addressesHex);
+
+            var arenaSheet = sheets.GetSheet<ArenaSheet>();
+            if (!arenaSheet.TryGetValue(championshipId, out var arenaRow))
+            {
+                throw new SheetRowNotFoundException(nameof(ArenaSheet),
+                    $"championship Id : {championshipId}");
+            }
+
+            if (!arenaRow.TryGetRound(round, out var roundData))
+            {
+                throw new RoundNotFoundException(
+                    $"[{nameof(BattleArena)}] ChampionshipId({arenaRow.Id}) - round({round})");
+            }
+
+            if (!roundData.IsTheRoundOpened(context.BlockIndex))
+            {
+                throw new ThisArenaIsClosedException(
+                    $"{nameof(BattleArena)} : block index({context.BlockIndex}) - " +
+                    $"championshipId({roundData.Id}) - round({roundData.Round})");
+            }
+
+            var arenaParticipantsAdr = ArenaParticipants.DeriveAddress(roundData.Id, roundData.Round);
+            if (!states.TryGetArenaParticipants(arenaParticipantsAdr, out var arenaParticipants))
+            {
+                throw new ArenaParticipantsNotFoundException(
+                    $"[{nameof(BattleArena)}] ChampionshipId({roundData.Id}) - round({roundData.Round})");
+            }
+
+            if (!arenaParticipants.AvatarAddresses.Contains(myAvatarAddress))
+            {
+                throw new AddressNotFoundInArenaParticipantsException(
```

Comment:
When my avatar information is not contain after the start of the round, if we can join process inside `BattleArena`, client does not have to wait `JoinArena`  render, and network can reduce the Tx to 1 from 2.

Reply:
We can only execute `BatteArena` action by those who participate in The Arena. But we'll discuss it for performance. 🏃🏻‍♂️